add webhooks documentation (#117)
foxglove/docs: docs/13-webhooks/0-introduction.md

---
title: Introduction

> description: Webhooks notify your software when events happen within your Org.

Reviewer: capitalization of Org seems weird

---

import AccountRequiredHeader from "../../src/components/docs/icons/AccountRequiredHeader";

<AccountRequiredHeader badgeText="Closed Beta, contact us for access" />

> Webhooks notify your software when events happen within your Foxglove org. Webhooks help synchronize state by providing push notifications for these events.

Reviewer: I think we can condense these two sentences somehow. Also I think we can reword to emphasize that webhooks are "providing you the building blocks" and less to possibly suggest they are synchronizing anything - that part is up to the user. Maybe just omit "synchronize".

> You create a _webhook_ that causes _notifications_ about _events_ to be _delivered_ to your _webhook endpoint_. Notifications are delivered as HTTP(S) POST requests with a JSON body. Check the [Concepts](./1-concepts.md) page for more details on each.

Reviewer: I think the italicization here is too heavy. It would be better to replace with "See the concepts page for details" or merge/link these sections somehow.

---
title: Introduction
description: Webhooks notify your software when events happen within your Org.
---

import AccountRequiredHeader from "../../src/components/docs/icons/AccountRequiredHeader";

> <AccountRequiredHeader badgeText="Requires Enterprise plan" />

Reviewer 2: does it actually require enterprise plan? I thought they were available on all plans

Author: I updated to reflect that the feature will debut in closed beta.

> Webhooks notify your software when events happen within your Org. Webhooks are useful when building automated systems that consume data after it has been uploaded, or to synchronize state between your Org and your own internal services.

Reviewer 3: I think we need to use the "Foxglove org" or similar language when starting the intro to solidify that we are referring to events and changes occurring within the Foxglove environment. I was confused the first time I read this because I didn't make that immediate leap.

```suggestion
Webhooks notify your software when events happen within your Foxglove Org. Webhooks help synchronize state by providing push notifications for these events.
```

> You create a _webhook_ that causes _notifications_ about _events_ to be _delivered_ to your _webhook endpoint_. Notifications are delivered as HTTP(S) POST requests with a JSON payload.

Reviewer 3: ```suggestion
You create a _webhook_ that causes _notifications_ about _events_ to be _delivered_ to your _webhook endpoint_. Notifications are delivered as HTTP(S) POST requests with a JSON body.
```

### Webhooks

A webhook subscribes your endpoint to notifications for a set of event types.

> Webhooks can be disabled and re-enabled. When disabled, notifications will not be delivered to your API endpoint.

Reviewer 3: This should go elsewhere - IMO too soon in the intro process to discuss this nuance that does not matter. It breaks the flow and adds noise.

> When you create a webhook, you receive a random token. This token allows you to verify the origin of POST requests to your endpoint. See [Webhook Security](2-security.md) for details on how to secure your webhook endpoint.

Reviewer 3: Seems not intro material - maybe better for getting started page?

Admins can create, edit, and delete webhooks in [Org Settings](https://console.foxglove.dev/settings/webhooks/).

### Webhook events

> Webhook events are records of occurrences within Foxglove. A webhook event consists of a _timestamp_, an _event type_, and data about the event, specific to that type.

Reviewer 3: Here you say "Foxglove" whereas above you had "Org". I recommended "Foxglove org" but "Foxglove" could be fine above as well.

> When an event occurs, all webhooks that subscribe to that event type receive a notification containing event data.

Reviewer 3: Not sure this needs to be said. Or if it does maybe said some other way. I did not find it useful information.
```suggestion
```

The following event types are available:

> | Name                 | Description                                                                     | Additional keys                                                                       |

Reviewer 3: "Additional keys" sounds weird. Above you say "and data about the event, specific to that type". So maybe call it "data"? Or "Payload"?

| -------------------- | ------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- |
| `recording.imported` | Occurs when a recording finishes [importing](./importing-data).                 | `recordingId`: The recording ID.                                                      |
| `device.created`     | Occurs when a new [Device](./importing-data#add-a-device) is added to your Org. | `deviceId`: The device ID.                                                            |
| `event.created`      | Occurs when a new [Event](./events) is added to your Org.                       | `eventId`: The event ID.                                                              |
| `ping`               | Ping event for debugging webhook endpoint connectivity.                         | `webhookId`: The webhook being pinged. Only this webhook will receive a notification. |

### Notification

> A notification is what your webhook endpoint receives when an event occurs. Notifications are delivered as JSON-formatted POST requests to an HTTP(S) endpoint on the public internet.

Reviewer 3: Something with this wording sounds weird to me. Maybe it would be better combined with the "Delivery" section below? Happy to chat on this or think async and provide more feedback. Somehow I think it is nice to tie this with delivery even though technically they are not the same?

A notification is delivered in the following form:

#### Request Headers

| Key                    | Value                                                                    |
| ---------------------- | ------------------------------------------------------------------------ |
| `Content-Type`         | `application/json`                                                       |

> | `fg-webhook-signature` | a signature verifying the request body as originating from your webhook. |

Reviewer 3: ```suggestion
| `fg-webhook-signature` | A signature verifying the request body as originating from your webhook. |
```

Reviewer 3: What format is this signature? Not much is said about this signature or how to learn more.

#### Request Body

A JSON object with the following keys:

| Key                   | Type                        | Description                                                                                                                           |
| --------------------- | --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------- |
| `timestamp`           | RFC3339-formatted timestamp | When the event occurred                                                                                                               |

> | `deliveryAttemptedAt` | RFC3339-formatted timestamp | When this delivery was attempted                                                                                                      |

Reviewer 3: UTC time of the attempting server? Why does this matter to me?

| `webhookId`           | string                      | The ID of the webhook that notified your endpoint                                                                                     |
| `eventId`             | string                      | A unique ID for this event. This is not unique across all notifications, since more than one webhook can notify about the same event. |
| `payload`             | object                      | Data about the event.                                                                                                                 |

The `payload` attribute will be an object with a `type` attribute matching the event type and some additional keys depending on the event type. See [Webhook Events](#webhook-events) above for descriptions of those keys.

> ### Deliveries

Reviewer 3: As a user, to me this seems synonymous with Notification - maybe worth thinking about combining. Another idea is to keep this introduction page simpler and go into details on these on separate pages.

> We target at-least-once delivery semantics for webhook notifications. This means a notification can be delivered more than once. Admins may also manually retry failed deliveries. This means your application may receive a notification long after the event occurs.

Reviewer 3: This sentence says a lot about this semantic but doesn't say the important thing about which parts of the delivery help me with idempotency.
